packages/dcos-integration-test/extra/test_3dt.py

```diff
@@ -458,28 +458,63 @@ def test_3dt_bundle_create(cluster):
     # make sure the job is done, timeout is 5 sec, wait between retying is 1 sec
     status_url = '/system/health/v1/report/diagnostics/status/all'
 
-    # the job executing takes about 1 seconds per node. A static timeout does not work very well for every context
-    # this test can be run in. Let's calculate a timeout based on how many nodes in a cluster + 15% tolerance.
-    # the timeout would looks like: (len(all_nodes) + 15%) * 1000
-    timeout = len(cluster.masters + cluster.all_slaves) * 1000
-    tolerance = int(15 * timeout / 100.0)
-    job_timeout = timeout + tolerance
-    logging.info("Using timeout {}".format(job_timeout))
-
-    @retrying.retry(stop_max_delay=job_timeout, wait_fixed=1000)
+    # lets make sure we are progressing. We need at least 2 points to compare.
+    def is_job_progressing(job_progress):
+        progress_points = len(job_progress)
+        if progress_points <= 1:
+            raise Exception("job progress must have at least 2 data points")
+
+        # we expect the progress to mutate over time. If this does not happen, the job is probably stuck.
+        for progress_point in range(progress_points - 1):
+            if job_progress[progress_point] >= job_progress[progress_point + 1]:
+                return False
+        return True
+
+    class StaleJobException(Exception):
+        """Exception should be raised if the job is not progressing"""
+
+    # calculate the `wait_for_job` timeout. Let's give 15 seconds per node. Should be good enough.
+    timeout = len(cluster.masters + cluster.all_slaves) * 15000
+
+    # if we encounter StaleJobException we should fail right away.
+    @retrying.retry(wait_fixed=2000, stop_max_delay=timeout,
+                    retry_on_exception=lambda e: not isinstance(e, StaleJobException))
     def wait_for_job():
```

Comment:
The retry loop inside a retry loop is a bit odd for a "continue forever" loop.

Why not make the main retry loop retry every 2 seconds or so, and have a couple state variables it keeps

1. Last state change time (`datetime.now()` when it got the last state change or got no state)
2. The last state retrieved (initially none)

The retry loop then does
1. fetch progress and print it for debugging (stdout/stderr will get captured by py.test)
2. check if state is done, and if so exit success
3. check if state has changed, and if so store new state, update last timestamp, continue / iterate in retry loop again
4. check if `datetime.now() - last_state_change > timedelta(seconds=15)`, and if so, exit loop with error (no progress made in 15 seconds).